### DISCUSSION TITLE
support proprietary protocol

### DISCUSSION THREAD

**Discussion by vsevel** [Hyperfoil/Hyperfoil#498] [Q&A] [answered]
hello,
we are using a proprietary protocol in my company (some rpc on top of http) that goes way back. it is used only for java to java communication.
it is essentially a matter of marshalling a request into the body of the http request, and unmarshalling the response body to recreate the java objects.
out of curiosity, is there a way to create an extension where I could implement so that scenarios would refer to it, instead of using http.
is `PluginRunData` the way to go?

**Comment by willr3** [comment]
I'm not sure what it would take to add an extension for an additional protocol but I will ask around and we will see what it would take.

**Comment by lampajr** [comment, marked as answer]
Hi @vsevel, thanks a lot for reaching out! 🚀 

Correct me if I am wrong, you'd like to leverages the HTTP plugin with just some modification in how the request is marshalled and unmarshalled. 

In this case, I think the easier way that comes to my mind is to use Hyperfoil [extensions](https://hyperfoil.io/docs/extensions/) in order to create a custom step/action that you would use instead of the `httpRequest`. There is also a good example [here](https://hyperfoil.io/docs/getting-started/quickstart8/) on how to create custom components. Additionally you might need to add a custom `Handler` as well to properly unmarshall the response.

Taking the following from the doc I shared above:
```txt
There are several other integration points but Step:

* `io.hyperfoil.api.session.Action` is very similar to step, but it does not allow blocking. Implement Action.BuilderFactory to define new actions.

* `StatusHandler`, `HeaderHandler` and `BodyHandler` in io.hyperfoil.api.http package process different stages of HTTP response parsing. All these have BuilderFactory inner interface for you to implement.

* `io.hyperfoil.api.connection.Processor` performs later generic stages of response processing. As this interface is generic, there are two factories that you could use: `i.h.a.c.Request.ProcessorBuilderFactory` and `i.h.a.c.HttpRequest.ProcessorBuilderFactory`.
```
And there is also a maven plugin that can scaffold some of the code, e.g.,
```bash
mvn io.hyperfoil:hyperfoil-codegen-maven-plugin:skeleton \
    -Dskeleton.package=foo.bar -Dskeleton.name=myComponent -Dskeleton.type=step
```

You can also take a look at [HttpRequestStepBuilder](https://github.com/Hyperfoil/Hyperfoil/blob/master/http/src/main/java/io/hyperfoil/http/steps/HttpRequestStepBuilder.java) and [PrepareHttpRequestStep](https://github.com/Hyperfoil/Hyperfoil/blob/master/http/src/main/java/io/hyperfoil/http/steps/PrepareHttpRequestStep.java)

Otherwise, the alternative would be to create a custom `plugin`, similarly to what have been done for [HotRod](https://github.com/Hyperfoil/Hyperfoil/tree/master/hotrod) but I think this would take much more time as you'd need to manage the whole lifecycle/configuration for the new protocol (which might be what you want, don't know).

Hope this will help you somehow but let me know if you need further clarification 🙏🏻

**Reply by vsevel** [reply]
perfect thanks
